File: exercises/RB101_small_problems/easy_5/letter_swap.rb
# Solve using PEDAC

# --- Understand the problem:

# Given a string of words seperated by spaces, write a method that
# takes this string of words and returns a string in which the first
# and last letters of every word are swapped.

# Explicit:
#   - assume every word contains one letter
#   - assume string will always contain one word.
#   - assume string contains nothing but words and spaces.
#   - Input: string
#   - Output: new string

# Implicit:
#   - return string if contains one character
#   - words are case-sensitive

# --- Examples / Test Cases:

# swap('Oh what a wonderful day it is') == 'hO thaw a londerfuw yad ti si'
# swap('Abcde') == 'ebcdA'
# swap('a') == 'a'

# --- Data Structure + Algorithm

#   */Given an string of words/*
# - return string if contains one character
# - split string into individual words
# - loop through array and modify each word
#   - split word using #chars
#   - save first and last elements to a variable
#   - reassing first and last elements
# - append word into a new array
# - join array and return

def swap(str)
  return str if str.size == 1
  result = []

  str.split(' ').each do |word|
    word[0], word[-1] = word[-1], word[0]
    result << word
  end

  result.join(' ')
end

p swap('Oh what a wonderful day it is') == 'hO thaw a londerfuw yad ti si'
p swap('Abcde') == 'ebcdA'
p swap('a') == 'a'

File: exercises/JS210_Small_Problems/Easy_3/letterSwap.js
// const swap = str => {
//   let words = str.split(' ');

//   for (let i = 0; i < words.length; i++) {
//     if (words[i].length === 1) continue;

//     let first = words[i][0];
//     let middle = words[i].slice(1, -1);
//     let last = words[i].slice(-1)

//     words[i] = last + middle + first
//   }

//   return words.join(' ');
// }

// Using map

function swapFirstLastCharacters(word) {
  if (word.length === 1) {
    return word;
  }

  return word[word.length - 1] + word.slice(1, -1) + word[0];
}

function swap(str) {
  return str.split(' ').map(swapFirstLastCharacters).join(' ');
}

console.log(swap('Oh what a wonderful day it is'));  // "hO thaw a londerfuw yad ti si"
console.log(swap('Abcde'));                          // "ebcdA"
console.log(swap('a'));                              // "a"